Rebuilding Roles / Users / Permissions

level09 opened the pull request.

apodacaduron committed:
basic ui updates for system users table

apodacaduron committed:
update create users dialog to match figma design

apodacaduron committed:
add default to toggle button component

apodacaduron committed:
Tweak message when new user created

apodacaduron committed:
fix typo on save button

apodacaduron committed:
allow users to use main input for search

apodacaduron committed:
update edit user dialog

apodacaduron committed:
when selecting user role mark permissions as selected

apodacaduron committed:
allow user to deselect permissions if he has admin role

apodacaduron committed:
view user updates

apodacaduron committed:
basic add of change password dialog

apodacaduron committed:
improvements to form validations on change password dialog

apodacaduron committed:
pass change password evt to dialog and use edited item to check force_reset flag

apodacaduron committed:
add saving flag to create user button

apodacaduron committed:
add saving flag to change password dialog

apodacaduron committed:
tweak 2fa badge on drawer

apodacaduron committed:
add force password confirmation to users table

apodacaduron committed:
tweak translation

apodacaduron committed:
add basic filtering ui

apodacaduron committed:
copy confirmation dialog from form builder branch

apodacaduron committed:
refactor password reset to use new global confirm dialog

apodacaduron committed:
add confirmation dialog for global password reset

apodacaduron committed:
add confirmations dialog to logout user sessions

apodacaduron committed:
fix card rounded edges

apodacaduron committed:
hide active account switch if user edits it's own account

apodacaduron committed:
reuse existing function

apodacaduron committed:
if no role assigned show view only text

apodacaduron committed:
update revoke2fa to use dialog confirmation

apodacaduron committed:
disable toggle button if user is admin

apodacaduron committed:
tweak user preview table columns

apodacaduron committed:
adjustment to translations

apodacaduron committed:
add asterisk to change password page

apodacaduron committed:
remove root property since users page is already on root

apodacaduron committed:
tweak height of users table

apodacaduron committed:
remove .once modifier from logout session and force global password reset

apodacaduron committed:
remove extra 2fa button

apodacaduron committed:
remove debug logs

apodacaduron committed:
add new access roles and allow only selecting one sustem role

apodacaduron committed:
mark all access roles as checked if user is admin

apodacaduron committed:
replace title for name key

level09 committed:
Add filtering support to users API endpoint

  - Search across username, email, and name fields
  - Filter by account status (active/inactive)
  - Filter by 2FA status (enabled/disabled)
  - Multi-select filter by roles

Supports frontend user dashboard filter UI implementation

apodacaduron committed:
Change add user button to simply show save changes

apodacaduron committed:
Merge branch 'main' of https://github.com/sjacorg/bayanat into BYNT-1475-Implement-new-user-dashboard-design

apodacaduron committed:
Merge branch 'BYNT-1475-Implement-new-user-dashboard-design' of https://github.com/sjacorg/bayanat into BYNT-1475-Implement-new-user-dashboard-design

apodacaduron committed:
fix filters on users page so it actually makes a request to backend

apodacaduron committed:
make search clearable and remove useless vars

apodacaduron committed:
add view only role with id 0

level09 committed:
Enhance users API filtering capabilities

- Added support for filtering users without roles using role ID 0
- Updated logic to handle cases where users may have no roles or selected roles
- Fix twoFactor filter bug when passed as false

level09 committed:
Merge branch 'BYNT-1475-Implement-new-user-dashboard-design' of github.com:sjacorg/bayanat into BYNT-1475-Implement-new-user-dashboard-design

apodacaduron committed:
add view only role to filter dropdown and not user create/edit dialog

apodacaduron committed:
fix missing components error

apodacaduron committed:
Mark with a checkmark and primary color selected filters

apodacaduron committed:
make user dialog fullscreen and add new access level buttons

apodacaduron committed:
if user is admin dont allow to deselect access roles

apodacaduron committed:
render access level and role on table and drawer

apodacaduron committed:
access level is now computed on frontend

apodacaduron committed:
tweaks to compute access level on frontend

apodacaduron committed:
update system role to use btn-toggle, change icon size, and update read only to be view only text

apodacaduron committed:
hide permissions from view only role

level09 committed:
feat: add UserHistory model for tracking user revisions

Introduces a new SQL Alchemy model, UserHistory, to maintain a history of user revisions. This model includes fields for target user ID, user ID, and a JSON column for storing revision data. It also provides methods for serialization and restricted data access, enhancing the auditing capabilities of user changes.

level09 committed:
add create_revision method to User model for tracking changes

level09 committed:
Register UserHistory model in admin models for user revision tracking

level09 committed:
add API endpoint for user history retrieval

level09 committed:
add API endpoint to temporarily suspend user accounts

level09 committed:
add API endpoint to reactivate suspended user accounts + bug fixes

level09 committed:
add API endpoint to disable user accounts

level09 committed:
add API endpoint to enable user accounts

level09 committed:
Implement sql migration script to efactor system roles in database

- Renamed existing roles: 'Mod' to 'Moderator' and 'DA' to 'Analyst'.
- Introduced a new 'View' role with read-only permissions.
- Added rollback script for potential reversions.

level09 committed:
Refactor role generation in data_helpers.py

- Consolidated role creation logic into a loop for better maintainability.
- Updated roles to include 'Admin', 'Analyst', 'Moderator', and 'View' with a unified description.
- Removed redundant code for individual role checks and creation.

apodacaduron committed:
Merge branch 'BYNT-1475-Implement-new-user-dashboard-design' of https://github.com/sjacorg/bayanat into roles-permissions-users

apodacaduron committed:
remove access level logic and add basic user statuses

apodacaduron committed:
add new translations

apodacaduron committed:
tweaks to roles sections and two fa

level09 committed:
enforce system role requirement for login

- override User.is_active to require at least one role
- add login form validation to log and notify admins when user without role attempts login
- users without roles cannot login and admins are notified

level09 committed:
add database migration for system roles refactor

- rename Mod to Moderator and DA to Analyst
- add new View role with read-only permissions
- assign View role to active users without roles to prevent lockout

level09 committed:
update role names in decorators

rename 'Mod' to 'Moderator' and 'DA' to 'Analyst' in all role decorators

level09 committed:
update role names in templates

update all has_role() and roles_in() calls to use new role names (Moderator, Analyst)

apodacaduron committed:
fix typo

apodacaduron committed:
add missing key to loop and render chip if user has [] permissions

apodacaduron committed:
move mixin declaration to top

apodacaduron committed:
add reusable confirm dialog for account manipulation

apodacaduron committed:
remove payload constant

apodacaduron committed:
allow disabling accept button from prop

apodacaduron committed:
wire up password field and account manipulation requests

apodacaduron committed:
initial try to merge both system and access roles into one array

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

level09 committed:
Simplify is_active property to use bool for roles check

level09 committed:
Update user role names in test fixtures for clarity: changed 'DA' to 'Analyst' and 'Mod' to 'Moderator'.

level09 committed:
Update role names in test fixture to reflect current terminology: changed 'DA' to 'Analyst' and 'Mod' to 'Moderator'.

level09 committed:
Add 'deleted' field to User model serialization

apodacaduron committed:
replace older role names with new ones

apodacaduron committed:
Prevent user from deleting view role as it's a default field

apodacaduron committed:
if user is disabled hide suspended button

apodacaduron committed:
change user disabled rule in computed

apodacaduron committed:
tweak table and role display, remove deleted field before saving and update table record when refreshing user

apodacaduron committed:
remove default logic to view role

apodacaduron committed:
remove useless template

apodacaduron committed:
hide x icon on toggle button when editin an admin user with incomplete permissions

apodacaduron committed:
add show snack messages

apodacaduron committed:
if user has no permissions, or access role, or system role add a message on the user card component

apodacaduron committed:
remove deleted flag from changePassword method

apodacaduron committed:
add quick access if user has no roles or permissions

apodacaduron committed:
fix issue if user has no access roles yet

apodacaduron committed:
replace password to confirm logic with username instead

apodacaduron committed:
expose current user id

apodacaduron committed:
hide account manipulation buttons if it's from the currently signed in user

level09 committed:
Fix missing and_ import in user filtering logic

level09 committed:
Restrict setting active field during user creation only; existing users must use dedicated endpoints for status changes

level09 committed:
Remove active field from allowed attributes in user payload validator

apodacaduron committed:
remove active flag manipulation

level09 committed:
Set default active status to True for new users during creation, allowing explicit override

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

level09 committed:
Fix user test fixtures to properly handle foreign key dependencies

level09 committed:
Add proper cleanup of user dependencies in users fixture

level09 committed:
Remove active field from user test payload to match validation model

level09 committed:
Add UserStatus enum to define user account statuses

level09 committed:
Implement user account status field and update is_active logic to reflect new status system

level09 committed:
Add set_status method to manage user status with validation and side effects

level09 committed:
Update user model to replace active and deleted fields with a unified status field in user payloads

level09 committed:
Refactor user model to set status during creation, replacing active field logic with status synchronization

level09 committed:
Update user suspension logic to utilize UserStatus enum, ensuring proper handling of suspended and disabled states

level09 committed:
Refine user reactivation and suspension error handling to improve clarity and enforce status checks using UserStatus enum

level09 committed:
Refactor user disable logic to utilize UserStatus enum, replacing direct field manipulation with set_status method for improved clarity and consistency

level09 committed:
Enhance user reactivation process by clarifying status changes; now sets status to SUSPENDED instead of ACTIVE, requiring role reassignment for access restoration.

level09 committed:
Add unified endpoint for setting user status; includes validation and error handling for status changes, ensuring proper transitions between states.

level09 committed:
Add migration to introduce user status field, replacing active boolean with a status enum. Includes data migration, NOT NULL constraint, default value, and index creation for improved query performance.

level09 committed:
Update api_users endpoint to replace 'active' filter with 'status' filter using UserStatus enum for improved clarity and validation of account status.

apodacaduron committed:
update users page to use status key instead of the old active / deleted

apodacaduron committed:
fix username not being cleared on reject

apodacaduron committed:
update key to be active

apodacaduron committed:
update last remains of the old deleted flag

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

apodacaduron committed:
Merge branch 'main' of https://github.com/sjacorg/bayanat into roles-permissions-users

apodacaduron committed:
add missing semi colons reported by gh ai

apodacaduron committed:
add dialog missing translations

apodacaduron committed:
add more translations

level09 committed:
Remove role-stripping on user disable

level09 committed:
Update user status logic to default to SUSPENDED for users without roles

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

level09 committed:
Update enable endpoint to activate users with roles directly

Users with roles are now set to ACTIVE when enabled, removing the  unnecessary SUSPENDED intermediate step. Users without roles still go to SUSPENDED and require role assignment before activation

level09 committed:
Simplified the unified /status endpoint, removed the forced DISABLED → SUSPENDED → ACTIVE flow. Now it just uses set_status() which already has  role validation built in

apodacaduron committed:
if user has no roles and is suspended disable button and add tooltip saying that a user needs to get assigned a system role to allow it being reactivated

apodacaduron committed:
if password reset was requested still allow to change password just hide the reset checkbox

apodacaduron committed:
Parse number to boolean on disabled prop

apodacaduron committed:
tweak index selection

apodacaduron committed:
move actionSections from data to computed

apodacaduron committed:
make toggle button more readable

apodacaduron committed:
get rid of timeout on cleanup and use v-dialog afterLeave event

apodacaduron committed:
remove cleanup function since now it's automatically run on afterLeave

apodacaduron committed:
add null safe operators when confirm dialog is cleaned up

tarekio commented on enferno/admin/models/UserHistory.py:
```diff
@@ -0,0 +1,58 @@
+    user = db.relationship("User", foreign_keys=[user_id])
+
+    @property
+    def restricted_data(self):
```
there's no need for this.

tarekio reviewed (dismissed):
A few comments to be addressed.

level09 commented on enferno/admin/views.py:
```diff
@@ -3279,7 +3280,7 @@ def api_bulletins(validated_data: dict) -> Response:
 
 @admin.post("/api/bulletin/")
-@roles_accepted("Admin", "DA")
+@roles_accepted("Admin", "Analyst")
```
Same here, Not in the original requirements. Separate ticket please

tarekio replied:
This is fine, will add a new issue.

level09 commented on enferno/admin/views.py:
```diff
@@ -5297,6 +5386,233 @@ def api_user_force_reset_all() -> Response:
 
 
+@admin.post("/api/user/<int:id>/suspend")
+@roles_required("Admin")
```
This would require frontend and backend changes. Separate ticket please

level09 commented on enferno/migrations/20251105_225708_refactor_system_roles.sql:
```diff
@@ -0,0 +1,44 @@
+ON CONFLICT DO NOTHING;
+
+-- 3. Assign View role to active users without any roles
+INSERT INTO roles_users (user_id, role_id)
```
The intent was to prevent lockouts during upgrade. But if you prefer admins assign manually, I can remove it.

tarekio replied:
Yes better we leave it to admins to decide.

level09 reviewed (commented):
@apodacaduron  All good, just a small nit

tarekio commented on enferno/admin/views.py:
```diff
@@ -4853,24 +4854,110 @@ def api_locationhistory(locationid: t.id) -> Response:
     page = request.args.get("page", 1, int)
     per_page = request.args.get("per_page", PER_PAGE, int)
     q = request.args.get("q")
+    status = request.args.get("status")
```
That's fine.

apodacaduron committed:
remove redundant check and rename variable

level09 committed:
Updated user history API endpoint to require "Admin" role

level09 commented on enferno/admin/views.py:
```diff
@@ -4853,24 +4854,110 @@ def api_locationhistory(locationid: t.id) -> Response:
+
+
+@admin.route("/api/userhistory/<int:userid>")
+@require_view_history
```
Fair point. Updated

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

level09 committed:
Add password confirmation to reactivate and enable endpoints

level09 committed:
Simplify UserHistory model for admin-only access

level09 committed:
Remove auto-assignment of View role in migration

apodacaduron committed:
on reactivate or enable send password to confirm

level09 committed:
Add Moderator role to all Analyst endpoints

level09 committed:
Update tests for Moderator access to Analyst endpoints

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

tarekio committed:
Improve roles descriptions

apodacaduron committed:
make system role required by default

apodacaduron committed:
tweak validation to allow selecting 0 values

apodacaduron committed:
rename data analyst to analyst

apodacaduron committed:
improve logic of system role selection, becoming an admin selects all roles, when leaving the admin role we reset selection

apodacaduron committed:
fix errors when selecting admin

apodacaduron committed:
fix empty roles array

apodacaduron committed:
add label to access role on admin selection, bring back previous selection, and do not select access roles on admin selection

apodacaduron committed:
add new translation

apodacaduron committed:
add history button with tooltip on toolbar and emit showHistory evt

apodacaduron committed:
call open history function to display drawer

apodacaduron committed:
revision history drawer on users page

apodacaduron committed:
map history so it looks better on diff

level09 committed:
Add /api/users/assignable/ endpoint for task assignment

Dedicated endpoint returning only active users in compact format.  Designed for Moderators to get user lists for assignment tasks  without access to sensitive filters (2FA, roles, status).

level09 committed:
Update roles decorator for /api/users/ endpoint to require "Admin" role only

level09 committed:
Update user endpoint tests for role-based access changes

  - Moderator now gets 403 on /api/users/ (admin-only)
  - Add tests for /api/users/assignable/ endpoint (admin + moderator)

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

apodacaduron committed:
update /users endpoint for /assignable so moderators can keep loading users lists

apodacaduron committed:
parse user agent to display in sessions table device and browser

apodacaduron committed:
add round toggle button to user card

apodacaduron committed:
fix if statement to allow showing no changes text on first revision item

apodacaduron commented:
@level09 small note, seems like changing to disabled or suspending does not create new items in revision history, it also might be a good idea to also log password change/reset requests from the admin to the revision history too

level09 committed:
Fix Activity.create() call and type signature in forms.py

level09 committed:
Merge branch 'roles-permissions-users' of github.com:sjacorg/bayanat into roles-permissions-users

level09 committed:
Add revision history for user status changes

level09 committed:
Add activity logging for admin force password reset

apodacaduron committed:
add show password button to change_password.html

apodacaduron committed:
when globally forcing password reload page to show the change password prompt

apodacaduron committed:
refresh users list if account status is changed suspend or disable

apodacaduron committed:
Merge branch 'main' of https://github.com/sjacorg/bayanat into roles-permissions-users

apodacaduron committed:
remove display name from update user payload

level09 committed:
Merge remote-tracking branch 'origin/main' into roles-permissions-users

apodacaduron committed:
Merge branch 'main' of https://github.com/sjacorg/bayanat into roles-permissions-users

apodacaduron committed:
Merge branch 'main' of https://github.com/sjacorg/bayanat into roles-permissions-users

level09 committed:
Merge branch 'main' into roles-permissions-users

level09 committed:
Fix user status management and use set_status() consistently

apodacaduron committed:
replace assigned_to.active with assigned_to.status since active is no longer present on api response

apodacaduron committed:
Merge branch 'roles-permissions-users' of https://github.com/sjacorg/bayanat into roles-permissions-users

level09 committed:
Use to_compact() for assigned_to and first_peer_reviewer in list views

level09 committed:
Merge branch 'main' into roles-permissions-users

level09 committed:
Fix create_inactive_user fixture to set status field

The fixture only set the deprecated active flag but not the status
field, causing test_delete_user_endpoint to fail since the delete
endpoint checks user.status instead of user.active.
